File: ai/workflows/personalDetailsWorkflow.js
import { END, START, StateGraph } from '@langchain/langgraph';
import { BankingState } from '../graph/state/bankingState.js';
import { formatFinancialResponse } from '../assistant/responseFormatting.js';
import {
  createExecutedWorkflowResponse,
  createWorkflowResponse
} from '../contracts/assistantResponseContract.js';

const leverageDataNode = async (state, config) => {
  const services = config?.configurable?.services;
  const result = await services.profileService.getUserProfile({ userId: state.session.userId });
  const workflowResponse = createExecutedWorkflowResponse({
    operation: 'get_user_identity',
    result
  });

  return {
    ...state,
    workflow: { ...state.workflow, activeWorkflow: 'unknown', currentPhase: 'Leverage Data' },
    personalDetails: {
      ...state.personalDetails,
      userProfile: result
    },
    execution: workflowResponse.execution,
    workflowResponse,
    ui: { ...state.ui, message: '' }
  };
};

const returnResponseNode = async (state) => {
  const workflowResponse = createWorkflowResponse({
    ...state.workflowResponse,
    message: formatFinancialResponse('get_user_identity', state.execution.result, state.session.userLanguage)
  });

  return {
    ...state,
    workflow: { ...state.workflow, currentPhase: 'Return Response with Suggestions' },
    workflowResponse,
    ui: {
      ...state.ui,
      message: workflowResponse.message
    }
  };
};

const personalDetailsWorkflowGraph = new StateGraph(BankingState)
  .addNode('leverage_data', leverageDataNode)
  .addNode('return_response_with_suggestions', returnResponseNode)
  .addEdge(START, 'leverage_data')
  .addEdge('leverage_data', 'return_response_with_suggestions')
  .addEdge('return_response_with_suggestions', END)
  .compile();

export const runPersonalDetailsWorkflow = async ({ state, services }) => (
  personalDetailsWorkflowGraph.invoke(state, { configurable: { services } })
);

File: ai/workflows/personal_details_workflow.py
from ai.contracts.assistant_response_contract import create_executed_workflow_response, create_workflow_response
from ai.assistant.response_formatting import format_financial_response


async def run_personal_details_workflow(*, state: dict, services: dict) -> dict:
    profile_service = (services or {}).get("profileService")
    user_language = (state.get("session") or {}).get("userLanguage", "en")

    result = await profile_service.get_user_profile(user_id=(state.get("session") or {}).get("userId"))
    workflow_response = create_executed_workflow_response(operation="get_user_identity", result=result)

    state = {
        **state,
        "workflow": {**(state.get("workflow") or {}), "activeWorkflow": "unknown", "currentPhase": "Leverage Data"},
        "personalDetails": {**(state.get("personalDetails") or {}), "userProfile": result},
        "execution": workflow_response["execution"],
        "workflowResponse": workflow_response,
        "ui": {**(state.get("ui") or {}), "message": ""},
    }

    message = format_financial_response("get_user_identity", state["execution"]["result"], user_language)
    final_response = create_workflow_response(**{**workflow_response, "message": message})

    return {
        **state,
        "workflow": {**(state.get("workflow") or {}), "currentPhase": "Return Response with Suggestions"},
        "workflowResponse": final_response,
        "ui": {**(state.get("ui") or {}), "message": message},
    }
